src/middleware/persist.ts

```diff
@@ -0,0 +1,353 @@
+      ) =>
+        MaybePromise<EPersistentStorageValue["state"]>
+    merge?:
+      ( persistedState: EPersistentStorageValue["state"]
+      , currentState: EState
+      ) =>
+        EState
+  }
+
+interface EPersistentStorage
+  { getItem: (name: EPersistentStorageName) =>
+      MaybePromise<E.Serialized<EPersistentStorageValue> | null>
+  , setItem:
+      ( name: EPersistentStorageName
+      , value: E.Serialized<EPersistentStorageValue>
+      ) =>
+        MaybePromise<void>
+  , removeItem: (name: EPersistentStorageName) => MaybePromise<void>
+  }
+
+interface EPersistStore
+  { persist:
+      { setOptions: (options: O.Partial<EPersistOptions>) => void
+      , clearStorage: () => void
+      , rehydrate: () => Thenable<void>
+      , hasHydrated: () => boolean
+      , onHydrate: (listener: (state: EState) => void) => () => void
+      , onFinishHydration: (listener: (state: EState) => void) => () => void
+      }
+  }
+
+const persistImpl: EPersist = (stateInitializer, _options) => (parentSet, parentGet, parentStore) => {
+  let options = { ...defaultOptions, ..._options }
+
+  let persistentStorage = tryElse(options.getStorage, () => undefined)
+  let persistentStorageGetItem = () =>
+    persistentStorage!.getItem(options.name)
+  let persistentStorageSetItem = (serializedState: E.Serialized<EPersistentStorageValue>) =>
+    persistentStorage!.setItem(options.name, serializedState)
+  let persistentStorageRemoveItem = () =>
+    persistentStorage!.removeItem(options.name)
+  
+  if (!persistentStorage) {
+    return stateInitializer(
+      (...a) => {
+        console.warn(messages.unableToUpdateItem(options.name))
+        parentSet(...a);
+      },
+      parentGet, 
+      update(parentStore, "setState", setState => (...a) => {
+        console.warn(messages.unableToUpdateItem(options.name))
+        setState(...a);
+      })
+    )
+  }
+ 
+  const updatePersistentStorage = () =>
+    thenablify(options.serialize)({
+      state: options.partialize({ ...parentGet() }),
+      version: options.version
+    })
+    .then(persistentStorageSetItem)
+    .throw()
+  
+
+  let initialState = stateInitializer(
+    (...a) => {
+      parentSet(...a);
+      updatePersistentStorage();
+    },
+    parentGet, 
+    update(parentStore, "setState", setState => (...a) => {
+      setState(...a);
+      updatePersistentStorage();
+    })
+  )
+  let initialStateFromPersistentStorage: EState | undefined;
+
+  let hasHydrated = false
+  let hydrationEmitter = emitter(new Set<(state: EState) => void>())
+  let finishHydrationEmitter = emitter(new Set<(state: EState) => void>())
+
+  const hydrate = () => {
+    hasHydrated = false;
+    hydrationEmitter.emit(parentGet());
+    let postRehydrationCallback = options.onRehydrateStorage?.(parentGet()) || undefined
+
+    return thenablify(persistentStorageGetItem)()
+    .then(serialized => serialized !== null ? options.deserialize(serialized) : null)
```

Comment:
Previously we were checking if `serialized` is falsey instead of null, which is less correct as `serialized` could even be `""`